### ISSUE TITLE
The output of Gemma 3 4B for TensorRT and Transformers is not the same, even when using float32

### ISSUE THREAD

**Issue by Alireza3242** [NVIDIA/TensorRT-LLM#4815]
### System Info

A100
tensorrt_llm 0.20.0rc3

### Who can help?

@kaiyux

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I ran the Gemma 3 4B model using both TensorRT and Transformers, but I’m not getting identical outputs.

Serving with Transformers:
```
from transformers import AutoTokenizer, Gemma3ForCausalLM
ckpt = "/home/dev/ra_workspace/ra_workspace/gemma3_4b/"

model = Gemma3ForCausalLM.from_pretrained(
    ckpt, torch_dtype=torch.float32, device_map="auto", attn_implementation="eager"
)

tokenizer = AutoTokenizer.from_pretrained(ckpt)

messages = [
    {"role": "user", "content": [{"type": "text", "text": "سلام چه طوری؟"}]}
]

inputs = tokenizer.apply_chat_template(
    messages, add_generation_prompt=True,
    return_dict=True, return_tensors="pt",
).to(model.device)

generation = model.generate(**inputs, max_new_tokens=256, do_sample=False)
```
Serving with TensorRT-LLM:

In this case, we used version 0.20.0rc3. The model was converted and built using the following arguments:
```
"convert": {
  "model-dir": model_files + "/model",
  "output-model-dir": "/app/data/tllm_checkpoint",
  "dtype": "float",
  "ckpt-type": "hf"
},
"build": {
  "checkpoint_dir": "/app/data/tllm_checkpoint",
  "output_dir": "/app/model_repository/tensorrt_llm/1",
  "gemm_plugin": "disable",
  "max_batch_size": "32"
}
```
Serve-time arguments:
```
"args": {
  "triton_max_batch_size": "128",
  "tokenizer_dir": model_files + "/model",
  "decoupled_mode": "true",
  "engine_dir": "/app/model_repository/tensorrt_llm/1",
  "batch_scheduler_policy": "max_utilization",
  "max_tokens_in_paged_kv_cache": "16384",
  "postprocessing_instance_count": "1",
  "preprocessing_instance_count": "1",
  "bls_instance_count": "1",
  "add_special_tokens": "false",
  "triton_backend": "tensorrtllm",
  "max_queue_delay_microseconds": "1000000",
  "max_beam_width": "1",
  "batching_strategy": "inflight_fused_batching",
  "max_attention_window_size": "1024/,1024/,1024/,1024/,1024/,16384",
  "accumulate_tokens": "",
  "tensorrt_llm_model_name": "tensorrt_llm",
  "max_queue_size": "100000",
  "encoder_engine_dir": "",
  "exclude_input_in_output": "true",
  "encoder_input_features_data_type": "TYPE_BF16",
  "logits_datatype": "TYPE_FP32",
  "xgrammar_tokenizer_info_path": ""
}
```

```
message='<bos><start_of_turn>user\nسلام چه طوری؟<end_of_turn>\n<start_of_turn>model\n'
data = {
  "text_input": message,
  "temperature": 0.00,
  "top_k": 1,
  "max_tokens": 1024,
  "stream": False
}

url = 'http://0.0.0.0:8000/v2/models/ensemble/generate_stream'

response = requests.post(url, json=data, headers={"Content-Type": "application/json"})

```

### Expected behavior

exact output

### actual behavior

they are not same as each other

### additional notes

We have modified some parts of the tensorrt_llm library to make it behave more similarly to transformers. The following changes were made:

1.
The value of eps in the layer_norm within the attention layer was not the same.
The eps value should be changed to 1e-6.

2.
In the file /usr/local/lib/python3.12/dist-packages/tensorrt_llm/functional.py,
inside the function:
```
def rms_norm(input: Tensor,
             normalized_shape: Union[int, Tuple[int]],
             num_groups: int = 1,
             weight: Optional[Tensor] = None,
             eps: float = 1e-06) -> Tensor:
```

We modified the section related to with precision("float32") as follows:

```
with precision("float32"):
    input_dtype = input.dtype
    fp32_input = cast(input, "float32")
    varx = pow(fp32_input, 2.0)

    varx = varx.mean(dim=dim, keepdim=True)
    denom = varx + eps
    denom = denom.sqrt()
    fp32_y = fp32_input / denom
    if num_groups > 1:
        fp32_y = fp32_y.view(old_shape)

    if weight is not None:
        fp32_y = fp32_y * weight
    y = cast(fp32_y, input_dtype)
```
3-
In the file /usr/local/lib/python3.12/dist-packages/tensorrt_llm/models/gemma/convert.py,
in the following section:
```
elif any(keyword in name for keyword in (
        "pre_attention_norm.scale",
        "pre_ffw_norm.scale",
        "final_norm.scale",
        "pre_attention_norm/vars/0",
        "pre_ffw_norm/vars/0",
        "rms_normalization/vars/0",
        "input_layernorm",
        "post_attention_layernorm",
        "pre_feedforward_layernorm",
        "post_feedforward_layernorm",
        "model.norm.weight",
        "q_norm.weight",
        "k_norm.weight",
)):
    param = param + 1.0  # upcasted to float32 in case of bfloat16
    add_trt_llm_weight(weights, trt_llm_name, param, 'float32')
```
The value of trt_llm_config.dtype was changed to "float32".

4-
In /usr/local/lib/python3.12/dist-packages/tensorrt_llm/models/gemma/model.py,
we changed:
```
self.input_layernorm = RmsNorm(normalized_shape=config.hidden_size,
                               eps=config.norm_epsilon,
                               dtype=config.dtype)
```
to:
```
self.input_layernorm = RmsNorm(normalized_shape=config.hidden_size,
                               eps=config.norm_epsilon,
                               dtype="float32")
```
However, despite all these changes, the outputs are still different.

**Comment by VALLIS-NERIA** [comment]
@brb-nv Can you take a look at this Gemma 3 issue?
@Alireza3242 We just added Gemma 3 support of pytorch backend (at main branch), can you please give a try?

**Comment by brb-nv** [comment]
Hi @Alireza3242, we added support mainly for `Gemma-3-1b-it` and I remember the outputs of that were very close (if not identical) to reference HF implementation with greedy sampling. That said, we're working on adding support for other Gemma3 variants - 4b, 12b, 27b. So, fixing these discrepancies will help those efforts as well.
- Can you please point me to the attention layernorm you're referring to? Are you referring to the `q_norm`, `k_norm` from [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/models/modeling_gemma3.py#L62-L67)? If yes, we're using the eps value from config and not hardcoding as such. So, unsure where the discrepancy is coming from.

**Comment by Alireza3242** [comment]
@brb-nv 
This eps:
https://github.com/NVIDIA/TensorRT-LLM/blob/039f7e311830b257b425141f9cbc644bbfb8a29b/tensorrt_llm/layers/attention.py#L369

**Comment by brb-nv** [comment]
Noticed a discrepancy in Gemma3-4b-it.
Current decoder implementation (which focuses on Gemma3-1b-it) doesn't have any rope scaling. This is indeed the case for Gemma3-1b-it [[link](https://huggingface.co/google/gemma-3-1b-it/blob/main/config.json#L29)]. Gemma3-4b-it, on the other hand, has rope scaling [[link](https://huggingface.co/google/gemma-3-4b-it/blob/main/config.json#L20-L23)].

**Comment by liyi-xia** [comment]
@Alireza3242 Hi, have you figured out the issue? I am also facing this issue. When I look at intermediate layer result, it is more or less close. But the final logits have diff around 0.02.

**Comment by liyi-xia** [comment]
I found two methods to reduce diff
1. Configure correct position embedding accordly
2. Change all norm layer to be float32

**Comment by brb-nv** [comment]
One other fix: https://github.com/NVIDIA/TensorRT-LLM/pull/5773
This is specific to Gemma3 VLMs which use rope scaling.

**Comment by brb-nv** [comment]
Please create LLM object using FlashInfer attention backend like this:
https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/defs/accuracy/test_llm_api_pytorch.py#L939-L944

**Comment by brb-nv** [comment]
Hi @geraldstanje, let's stick to this instance of Github issue so that we anyone else exploring Gemma3 has full trial of information here. Currently, there are too many and I'm closing others with ref to this one:
https://github.com/NVIDIA/TensorRT-LLM/issues/6193#issuecomment-3236959868
https://github.com/NVIDIA/TensorRT-LLM/issues/4825#issuecomment-3233172791
https://github.com/NVIDIA/TensorRT-LLM/issues/7339#issuecomment-3233157207

I don't anticipate format of messages to be any different from what you'd send to a decoder-only LLM. 

Sorry, I don't have have much experience with Triton Inference Server. Instructions to use Gemma3 VLM with trtllm-serve can be found here:
https://github.com/NVIDIA/TensorRT-LLM/tree/release/1.0/examples/models/core/multimodal#gemma3

**Comment by geraldstanje1** [comment]
@brb-nv any eta when Gemma 3 4B will be ready with tensorrt llm with trtllm-build etc.?

**Comment by JoJoLev** [comment]
I'd also like info on Gemma 3 4b with tensorrt llm trtllm-build setup since I'm doing text input and would like to deploy on triton.

**Comment by brb-nv** [comment]
I don't think we'll extend beyond existing support for Gemma3 in TRT workflow. We're prioritizing Pytorch workflow going forward for all models.

**Comment by JoJoLev** [comment]
@brb-nv is there ability to run on Nvidia triton with the PyTorch workflow?

**Comment by JoJoLev** [comment]
@geraldstanje I got tensorrt-LLM running on triton with the torch backend recently.

**Comment by geraldstanje1** [comment]
hi @JoJoLev can you show an full running example?

**Comment by JoJoLev** [comment]
@geraldstanje1 sure!
I'll share here shortly.
I got the gemma3-4b running on triton with the torch backend for tensorrt-LLM.
It isn't as performant as I'd want it to be but I believe this could be due to some config settings, please tinker with the configs and lmk if you get something better!

**Comment by geraldstanje1** [comment]
@JoJoLev do you run this with nvcr.io/nvidia/tritonserver:25.08-trtllm-python-py3 ?

**Comment by geraldstanje** [comment]
@JoJoLev did you try enable it witch batching?

**Comment by JoJoLev** [comment]
@geraldstanje 
@geraldstanje1 

No, I just used those base values. Were you able to change any config to improve performance?

**Comment by JoJoLev** [comment]
@geraldstanje yes, I will probably quantize this one. Not sure if we are supposed to expect the same perf between the torch backend and the tensorrtllm core engine compilation path though? That doesn't seem very clear to me.

**Comment by karljang** [comment]
Sorry we never replied to this one. Closing as stale since the TensorRT engine-build path is deprecated in favor of the PyTorch backend. Please open a new issue if the problem persists in the latest version.
